README.md
# gcs-browser-upload [ ![Codeship Status for qubitdigital/gcs-browser-upload](https://codeship.com/projects/a9a3f9a0-b3b0-0137-978c-1a359acd0e85/status?branch=master)](https://codeship.com/projects/363673)

Chunked, pausable, recoverable uploading to Google Cloud Storage directly from the browser.


## How does it work?

1. User selects a file
1. File + a [Google Cloud Storage resumable upload URL](https://cloud.google.com/storage/docs/json_api/v1/how-tos/upload#resumable) are given to `gcs-browser-upload`
1. File is read in chunks
1. A checksum of each chunk is stored in `localStorage` once succesfully uploaded
1. If the page is closed and re-opened for some reason, the upload can be resumed by passing the same file and URL back to `gcs-browser-upload`. The file will be validated against the stored chunk checksums to work out if the file is the same and where to resume from.
1. Once the resume index has been found, `gcs-browser-upload` will continue uploading from where it left off.
1. At any time, the `pause` method can be called to delay uploading the remaining chunks. The current chunk will be finished. `unpause` can then be used to continue uploading the remaining chunks.


## Example

```js
import Upload from 'gcs-browser-upload'

let input = document.getElementById('fileInput')
let pause = document.getElementById('pause')
let unpause = document.getElementById('unpause')
let upload = null

input.addEventListener('change', async () => {
  upload = new Upload({
    id: 'foo',
    url: 'https://www.googleapis.com/..../....',
    file: input.files[0],
    onChunkUpload: (info) => {
      console.log('Chunk uploaded', info)
    }
  })

  try {
    await upload.start()
    console.log('Upload complete!')
  } catch (e) {
    console.log('Upload failed!', e)
  } finally () {
    upload = null
  }
})

pause.addEventListener('click', () => {
  if (upload) {
    upload.pause()
  }
})

unpause.addEventListener('click', () => {
  if (upload) {
    upload.unpause()
  }
})
```


## Config

```js
{
  id: null, // required - a unique ID for the upload 
  url: null, // required - GCS resumable URL
  file: null, // required - instance of File
  chunkSize: 262144, // optional - chunk size must be a multiple of 262144
  storage: window.localStorage, // optional - storage mechanism used to persist chunk meta data
  contentType: 'text/plain', // optional - content type of the file being uploaded
  onChunkUpload: () => {} // optional - a function that will be called with progress information
}
```


## Requirements

This library requires `regeneratorRuntime` to be available globally - it is written in ES7 and makes use of async/await, which gets compiled into generators. You can find out about regenerator-runtime [here](https://www.npmjs.com/package/regenerator-runtime).

## Handling errors

Various errors are thrown if something goes wrong during uploading. See [src/errors.js](https://github.com/qubitdigital/gcs-browser-upload/blob/master/src/errors.js) for the different types. These are exported as a property on the `Upload` class.


## Developing

```js
make bootstrap     // install dependencies
make test          // run tests
make test-watch    // continuously run tests
```


CHANGELOG.md
# Changelog

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
## [1.0.6] - 2019-10-01
### Fixes
- Upgrades axios and mocha to fix security issues.

## [1.0.5] - 2019-09-29
### Added
- Ensures successfull file transmision by checking GCS response headers containing chunk and file checksums,
  re-transmitting chunks when they differ.

### Fixed
- SparkMD5 states was being saved/restored without cloning the structure, causing the values to be changed in unexpected
  ways and giving wrong checksum values.

## [1.0.4] - 2019-09-14
### Fixed
- The exponential backoff changed a bit the upload semantics, breaking the tests. This release maintains the pre-1.0.3
  semantics, making all green again!

## [1.0.3] - 2019-09-14
### Added
- Exponential backoff to retry failed chunk transmissions.
- Store SparkMD5 state, together with the chunk checksum, to localStorage so we can continue to calculate the file
  checksum without having to start from scratch. This also corrects a bug: the second time you started a resumable,
  a wrong checksum is calculated for the chunks uploaded in the first resume operation, discarding all the upload
  progress and starting from scratch.

### Changed
- Increase the transmission chunk size to 5mb.
- Only checks the two last uploaded chunks (10mb) before starting a resume operation.

[Unreleased]: https://github.com/fnix/gcs-browser-upload/compare/v1.0.6...HEAD
[1.0.6]: https://github.com/fnix/gcs-browser-upload/compare/v1.0.5...v1.0.6
[1.0.5]: https://github.com/fnix/gcs-browser-upload/compare/v1.0.4...v1.0.5
[1.0.4]: https://github.com/fnix/gcs-browser-upload/compare/v1.0.3...v1.0.4
[1.0.3]: https://github.com/fnix/gcs-browser-upload/compare/1e4600fb4f117a6f997a3162a039e28e9686cf24...v1.0.3


test/lib/server.js
import axios from 'axios'
import express from 'express'
import getPort from 'get-port'
import pify from 'pify'
import bodyParser from 'body-parser'

let server = null
let requests = []
let file = null
const router = new express.Router()

router.use(bodyParser.text())

router.use((req, res, next) => {
  const range = req.headers['content-range']
  const matchKnown = range.match(/^bytes (\d+?)-(\d+?)\/(\d+?)$/)
  const matchUnknown = range.match(/^bytes \*\/(\d+?)$/)

  if (matchUnknown) {
    req.range = {
      known: false,
      total: matchUnknown[1]
    }
    next()
  } else if (matchKnown) {
    req.range = {
      known: true,
      start: matchKnown[1],
      end: matchKnown[2],
      total: matchKnown[3]
    }
    next()
  } else {
    res.status(400).send('No valid content-range header provided')
  }
})

router.use((req, res, next) => {
  requests.push({
    method: req.method,
    url: req.originalUrl,
    headers: req.headers,
    body: req.body
  })
  next()
})

router.put('/', (req, res) => {
  if (!file) {
    file = {
      total: req.range.total,
      index: 0
    }
  }

  if (req.range.known) {
    file.index = req.range.end
  }
  res.set('range', `bytes=0-${file.index}`)
  if (file.index + 1 === file.total) {
    res.send(200).send('OK')
  } else {
    res.status(308).send('Resume Incomplete')
  }
})

router.put('/fail', (req, res) => {
  res.status(500).send('Internal Server Error')
})

export async function start () {
  const port = await getPort()
  const app = express()
  app.use('/file', router)
  server = await pify(::app.listen)(port)
  axios.defaults.baseURL = `http://localhost:${port}`
  axios.defaults.validateStatus = false
}

export function resetServer () {
  requests = []
  file = null
}

export function stop () {
  if (server) {
    server.close()
    server = null
  }
}

export function getRequests () {
  return requests
}


test/lib/makeFile.js
export default function makeFile (data) {
  return {
    size: data.length,
    name: 'foo',
    buffer: Buffer.from(data),
    slice: (start, end) => makeFile(data.substring(start, end))
  }
}


test/functional-tests.js
import localStorage from 'localStorage'
import { expect } from 'chai'
import randomString from 'random-string'
import Upload from '../src/Upload'
import { start, resetServer, stop, getRequests } from './lib/server'
import makeFile from './lib/makeFile'
import waitFor from './lib/waitFor'

describe('Functional', () => {
  before(start)
  after(stop)

  let upload = null
  let file = null
  let requests = []

  async function doUpload (length, url, retries = 2) {
    if (length !== null) {
      file = randomString({ length })
    }
    upload = new Upload({
      id: 'foo',
      url: url || '/file',
      chunkSize: 256,
      file: makeFile(file),
      retries: retries
    }, true)
    await upload.start()
    requests = getRequests()
    return upload
  }

  function reset () {
    localStorage.clear()
    resetServer()
    if (upload) {
      upload.cancel()
      upload = null
    }
  }

  describe('a single-chunk upload', () => {
    before(() => doUpload(256))
    after(reset)

    it('should only upload one chunk', () => {
      expect(requests).to.have.length(1)
    })

    it('should make a PUT request to the right URL', () => {
      expect(requests[0].method).to.equal('PUT')
      expect(requests[0].url).to.equal('/file')
    })

    it('should send the correct headers', () => {
      expect(requests[0].headers).to.containSubset({
        'content-length': '256',
        'content-range': 'bytes 0-255/256'
      })
    })

    it('should send the file in the body', () => {
      expect(requests[0].body).to.equal(file)
    })
  })

  describe('a multi-chunk upload', () => {
    before(() => doUpload(700))
    after(reset)

    it('should upload multiple chunks', () => {
      expect(requests).to.have.length(3)
    })

    it('should make multiple PUT requests to the right URL', () => {
      requests.forEach((request) => {
        expect(request.method).to.equal('PUT')
        expect(request.url).to.equal('/file')
      })
    })

    it('should send the correct headers', () => {
      expect(requests[0].headers).to.containSubset({
        'content-length': '256',
        'content-range': 'bytes 0-255/700'
      })
      expect(requests[1].headers).to.containSubset({
        'content-length': '256',
        'content-range': 'bytes 256-511/700'
      })
      expect(requests[2].headers).to.containSubset({
        'content-length': '188',
        'content-range': 'bytes 512-699/700'
      })
    })

    it('should send a total content length identical to the upload file size', () => {
      const totalSize = requests.reduce((result, request) => {
        return result + parseInt(request.headers['content-length'])
      }, 0)
      expect(totalSize).to.equal(700)
    })

    it('should send the file in the body', () => {
      expect(requests[0].body).to.equal(file.substring(0, 256))
      expect(requests[1].body).to.equal(file.substring(256, 512))
      expect(requests[2].body).to.equal(file.substring(512, 701))
    })
  })

  describe('a paused then resumed upload', () => {
    after(reset)

    it('should stop uploading after being paused', async () => {
      doUpload(500)
      upload.pause()
      await waitFor(() => {
        requests = getRequests()
        return requests.length > 0
      })
      expect(requests).to.have.length(1)
      expect(requests[0].body).to.equal(file.substring(0, 256))
    })

    it('should check the server for status before resuming', async () => {
      await doUpload(null)
      expect(requests[1].body).to.deep.equal({})
    })

    it('should send the rest of the chunks after being resumed', () => {
      expect(requests).to.have.length(3)
      expect(requests[2].body).to.equal(file.substring(256, 501))
    })
  })

  describe('a paused upload that is resumed with a different file', () => {
    let firstFile = null

    before(async () => {
      doUpload(500)
      firstFile = file
      upload.pause()
      await waitFor(() => getRequests().length > 0)
    })
    after(reset)

    it('should check the server for status before resuming', async () => {
      await doUpload(500)
      expect(requests[1].body).to.deep.equal({})
    })

    it('should upload part of the first file', () => {
      expect(firstFile).to.not.equal(file)
      expect(requests[0].body).to.equal(firstFile.substring(0, 256))
    })

    it('should upload the entire new file', () => {
      expect(requests).to.have.length(4)
      expect(requests[2].body).to.equal(file.substring(0, 256))
      expect(requests[3].body).to.equal(file.substring(256, 501))
    })
  })

  describe('an upload to a url that doesn\'t exist', () => {
    it('should throw a UrlNotFoundError', () => {
      return expect(doUpload(200, '/notfound')).to.be.rejectedWith(Upload.errors.UrlNotFoundError)
    })
  })

  describe('an upload that results in a server error', () => {
    it('should throw an UploadFailedError', () => {
      return expect(doUpload(200, '/file/fail', 0)).to.be.rejectedWith(Upload.errors.UploadFailedError)
    })
  })
})


src/FileMeta.js
const STORAGE_KEY = '__gcsBrowserUpload'

export default class FileMeta {
  constructor (id, fileSize, chunkSize, storage) {
    this.id = id
    this.fileSize = fileSize
    this.chunkSize = chunkSize
    this.storage = storage
  }

  getMeta () {
    const meta = this.storage.getItem(`${STORAGE_KEY}.${this.id}`)
    if (meta) {
      return JSON.parse(meta)
    } else {
      return {
        checksums: [],
        chunkSize: this.chunkSize,
        started: false,
        fileSize: this.fileSize
      }
    }
  }

  setMeta (meta) {
    const key = `${STORAGE_KEY}.${this.id}`
    if (meta) {
      this.storage.setItem(key, JSON.stringify(meta))
    } else {
      this.storage.removeItem(key)
    }
  }

  isResumable () {
    let meta = this.getMeta()
    return meta.started && this.chunkSize === meta.chunkSize
  }

  getResumeIndex () {
    return this.getMeta().checksums.length
  }

  getFileSize () {
    return this.getMeta().fileSize
  }

  addChecksum (index, checksum, state) {
    let meta = this.getMeta()
    meta.checksums[index] = { checksum: checksum, state: state }
    meta.started = true
    this.setMeta(meta)
  }

  getChecksum (index) {
    return this.getMeta().checksums[index].checksum
  }

  getSparkMD5State (index) {
    return JSON.parse(this.getMeta().checksums[index].state)
  }

  reset () {
    this.setMeta(null)
  }
}


src/FileProcessor.js
import { Promise } from 'es6-promise'
import SparkMD5 from 'spark-md5'
import debug from './debug'

class FileProcessor {
  constructor (file, meta, chunkSize) {
    this.paused = false
    this.file = file
    this.meta = meta
    this.chunkSize = chunkSize
    this.unpauseHandlers = []
  }

  async run (fn, startIndex = 0, endIndex) {
    const { file, chunkSize } = this
    const totalChunks = Math.ceil(file.size / chunkSize)
    let spark = new SparkMD5.ArrayBuffer()
    if (startIndex > 0) {
      debug('Restoring SparkMD5 state for the last chunck')
      spark.setState(this.meta.getSparkMD5State(startIndex - 1))
    }

    debug('Starting run on file:')
    debug(` - Total chunks: ${totalChunks}`)
    debug(` - Start index: ${startIndex}`)
    debug(` - End index: ${endIndex || totalChunks}`)

    const processIndex = async (index) => {
      if (index === totalChunks || index === endIndex) {
        debug('File process complete')
        return
      }
      if (this.paused) {
        await waitForUnpause()
      }

      const start = index * chunkSize
      const section = file.slice(start, start + chunkSize)
      const chunk = await getData(file, section)
      const { checksum, state } = computeChecksum(spark, chunk)

      const shouldContinue = await fn(checksum, state, index, chunk)
      if (shouldContinue !== false) {
        await processIndex(index + 1)
      }
    }

    const waitForUnpause = () => {
      return new Promise((resolve) => {
        this.unpauseHandlers.push(resolve)
      })
    }

    await processIndex(startIndex)
  }

  pause () {
    this.paused = true
  }

  unpause () {
    this.paused = false
    this.unpauseHandlers.forEach((fn) => fn())
    this.unpauseHandlers = []
  }
}

function computeChecksum (spark, chunk) {
  spark.append(chunk)
  const state = JSON.stringify(spark.getState())
  const checksum = spark.end()
  spark.setState(JSON.parse(state))
  return { checksum, state }
}

async function getData (file, blob) {
  return new Promise((resolve, reject) => {
    let reader = new window.FileReader()
    reader.onload = () => resolve(reader.result)
    reader.onerror = reject
    reader.readAsArrayBuffer(blob)
  })
}

export default FileProcessor


src/Upload.js
import { put } from 'axios'
import pRetry from 'p-retry'

import FileMeta from './FileMeta'
import FileProcessor from './FileProcessor'
import debug from './debug'
import {
  DifferentChunkError,
  FileAlreadyUploadedError,
  UrlNotFoundError,
  UploadFailedError,
  UnknownResponseError,
  MissingOptionsError,
  UploadIncompleteError,
  InvalidChunkSizeError,
  UploadAlreadyFinishedError
} from './errors'
import * as errors from './errors'

const MIN_CHUNK_SIZE = 5242880 // 5mb

export default class Upload {
  static errors = errors;

  constructor (args, allowSmallChunks) {
    var opts = {
      chunkSize: MIN_CHUNK_SIZE,
      storage: window.localStorage,
      contentType: 'text/plain',
      onChunkUpload: () => { },
      id: null,
      url: null,
      file: null,
      retries: 10,
      ...args
    }

    if ((opts.chunkSize % MIN_CHUNK_SIZE !== 0 || opts.chunkSize === 0) && !allowSmallChunks) {
      throw new InvalidChunkSizeError(opts.chunkSize)
    }

    if (!opts.id || !opts.url || !opts.file) {
      throw new MissingOptionsError()
    }

    debug('Creating new upload instance:')
    debug(` - Url: ${opts.url}`)
    debug(` - Id: ${opts.id}`)
    debug(` - File size: ${opts.file.size}`)
    debug(` - Chunk size: ${opts.chunkSize}`)

    this.opts = opts
    this.meta = new FileMeta(opts.id, opts.file.size, opts.chunkSize, opts.storage)
    this.processor = new FileProcessor(opts.file, this.meta, opts.chunkSize)
    this.lastResult = null
  }

  async start () {
    const { meta, processor, opts, finished } = this

    const resumeUpload = async () => {
      const localResumeIndex = meta.getResumeIndex()
      const remoteResumeIndex = await getRemoteResumeIndex()

      const resumeIndex = Math.min(localResumeIndex, remoteResumeIndex)
      debug(`Validating chunks up to index ${resumeIndex}`)
      debug(` - Remote index: ${remoteResumeIndex}`)
      debug(` - Local index: ${localResumeIndex}`)

      try {
        let startResumeIndex = resumeIndex - 2
        if (startResumeIndex < 0) startResumeIndex = 0

        await processor.run(validateChunk, startResumeIndex, resumeIndex)
      } catch (e) {
        debug('Validation failed, starting from scratch')
        debug(` - Failed chunk index: ${e.chunkIndex}`)
        debug(` - Old checksum: ${e.originalChecksum}`)
        debug(` - New checksum: ${e.newChecksum}`)

        await processor.run(uploadChunk)
        return
      }

      debug('Validation passed, resuming upload')
      await processor.run(uploadChunk, resumeIndex)
    }

    const uploadChunk = async (checksum, state, index, chunk) => {
      const total = opts.file.size
      const start = index * opts.chunkSize
      const end = index * opts.chunkSize + chunk.byteLength - 1

      const options = {
        headers: {
          'Content-Type': opts.contentType,
          'Content-Range': `bytes ${start}-${end}/${total}`
        },
        onUploadProgress: (progressEvent) => {
          opts.onChunkUpload({
            totalBytes: total,
            uploadedBytes: start + progressEvent.loaded,
            chunkIndex: index,
            chunkLength: chunk.byteLength
          })
        },
        validateStatus: false
      }

      debug(`Uploading chunk ${index}:`)
      debug(` - Chunk length: ${chunk.byteLength}`)
      debug(` - Start: ${start}`)
      debug(` - End: ${end}`)

      const res = await pRetry(async () => {
        const current_res = await safePut(opts.url, chunk, options)

        try {
          checkResponseStatus(current_res, opts, [200, 201, 308])
          checkResponseHeaders(current_res, { index, end, checksum })
        } catch (e) {
          if (e instanceof UrlNotFoundError) {
            throw new pRetry.AbortError(e)
          } else {
            throw e
          }
        }

        return current_res
      }, { retries: opts.retries })
      this.lastResult = res
      debug(`Chunk upload succeeded, adding checksum ${checksum}`)
      meta.addChecksum(index, checksum, state)

      opts.onChunkUpload({
        totalBytes: total,
        uploadedBytes: end + 1,
        chunkIndex: index,
        chunkLength: chunk.byteLength
      })
    }

    const validateChunk = async (newChecksum, _state, index) => {
      const originalChecksum = meta.getChecksum(index)
      const isChunkValid = originalChecksum === newChecksum
      if (!isChunkValid) {
        meta.reset()
        throw new DifferentChunkError(index, originalChecksum, newChecksum)
      }
    }

    const getRemoteResumeIndex = async () => {
      const options = {
        headers: {
          'Content-Range': `bytes */${opts.file.size}`
        },
        validateStatus: false
      }
      debug('Retrieving upload status from GCS')
      const res = await safePut(opts.url, null, options)

      checkResponseStatus(res, opts, [308])
      const header = res.headers['range']
      debug(`Received upload status from GCS: ${header}`)
      const range = header.match(/(\d+?)-(\d+?)$/)
      const bytesReceived = parseInt(range[2]) + 1
      return Math.floor(bytesReceived / opts.chunkSize)
    }

    if (finished) {
      throw new UploadAlreadyFinishedError()
    }

    if (meta.isResumable() && meta.getFileSize() === opts.file.size) {
      debug('Upload might be resumable')
      await resumeUpload()
    } else {
      debug('Upload not resumable, starting from scratch')
      await processor.run(uploadChunk)
    }
    debug('Upload complete, resetting meta')
    meta.reset()
    this.finished = true
    return this.lastResult
  }

  pause () {
    this.processor.pause()
    debug('Upload paused')
  }

  unpause () {
    this.processor.unpause()
    debug('Upload unpaused')
  }

  cancel () {
    this.processor.pause()
    this.meta.reset()
    debug('Upload cancelled')
  }
}

function hexToBase64 (hexstring) {
  return btoa(hexstring.match(/\w{2}/g).map(function (a) {
    return String.fromCharCode(parseInt(a, 16))
  }).join(''))
}

function checkResponseStatus (res, opts, allowed = []) {
  const { status } = res
  if (allowed.indexOf(status) > -1) {
    return true
  }

  switch (status) {
    case 308:
      throw new UploadIncompleteError()

    case 201:
    case 200:
      throw new FileAlreadyUploadedError(opts.id, opts.url)

    case 404:
      throw new UrlNotFoundError(opts.url)

    case 500:
    case 502:
    case 503:
    case 504:
      throw new UploadFailedError(status)

    default:
      throw new UnknownResponseError(res)
  }
}

function checkResponseHeaders (res, chunkInfo) {
  const { headers } = res
  const receivedMD5 = headers['x-goog-hash'] || headers['x-range-md5']

  if (!receivedMD5) return

  if (headers['x-goog-hash'] &&
      (chunkInfo.end + 1) === headers['x-goog-stored-content-length'] &&
      receivedMD5.match(/md5=(.*)/)[1] !== hexToBase64(chunkInfo.checksum)) {
    throw new DifferentChunkError(chunkInfo.index, hexToBase64(chunkInfo.checksum), receivedMD5)
  } else if (headers['range']) {
    const range = headers['range'].match(/(\d+?)-(\d+?)$/)
    const bytesReceived = parseInt(range[2])

    if (bytesReceived === chunkInfo.end && receivedMD5 !== chunkInfo.checksum) {
      throw new DifferentChunkError(chunkInfo.index, chunkInfo.checksum, receivedMD5)
    }
  }
}

async function safePut () {
  try {
    return await put.apply(null, arguments)
  } catch (e) {
    if (e instanceof Error) {
      throw e
    } else {
      return e
    }
  }
}
